mongofiles/options_test.go

```diff
@@ -7,76 +7,64 @@
 package mongofiles
 
 import (
-	"fmt"
+	"github.com/mongodb/mongo-go-driver/mongo/writeconcern"
 	"testing"
 
-	"github.com/mongodb/mongo-tools/common/db"
-	"github.com/mongodb/mongo-tools/common/options"
 	"github.com/mongodb/mongo-tools/common/testtype"
 	. "github.com/smartystreets/goconvey/convey"
 )
 
-// Regression test for TOOLS-1741
-func TestWriteConcernWithURIParsing(t *testing.T) {
+func TestWriteConcernOptionParsing(t *testing.T) {
 	testtype.SkipUnlessTestType(t, testtype.UnitTestType)
-	Convey("With an IngestOptions and ToolsOptions", t, func() {
+	Convey("Testing write concern parsing from command line and URI", t, func() {
+		Convey("Parsing with neither URI nor command line option should leave write concern empty", func() {
+			storageOpts := &StorageOptions{}
+			inputOpts := &InputOptions{}
+			_, opts, err := ParseOptions([]string{}, storageOpts, inputOpts)
 
-		// create an 'EnabledOptions' to determine what options should be able to be
-		// parsed and set form the input.
-		enabled := options.EnabledOptions{URI: true}
-
-		// create a new tools options to hold the parsed options
-		opts := options.New("", "", enabled)
+			So(err, ShouldBeNil)
+			So(storageOpts.WriteConcern, ShouldEqual, "")
+			So(opts.WriteConcern, ShouldBeNil)
+		})
 
-		// create a 'StorageOptions', which holds the value of the write concern
-		// for mongofiles.
-		storageOpts := &StorageOptions{}
-		opts.AddOptions(storageOpts)
+		Convey("Parsing with URI with no write concern specified in it should not error", func() {
+			args := []string{
+				"--uri", "mongodb://localhost:27017/test",
+			}
 
-		// Specify that a write concern set on the URI is not an error and is a known
-		// possible option.
-		opts.URI.AddKnownURIParameters(options.KnownURIOptionsWriteConcern)
+			storageOpts := &StorageOptions{}
+			inputOpts := &InputOptions{}
+			_, opts, err := ParseOptions(args, storageOpts, inputOpts)
 
-		Convey("Parsing with no value should leave write concern empty", func() {
-			_, err := opts.ParseArgs([]string{})
 			So(err, ShouldBeNil)
-			So(storageOpts.WriteConcern, ShouldEqual, "")
-			Convey("and building write concern object, WMode should be majority", func() {
```

Comment:
Related question: should we be explicitly setting majority as the default? Or should we be using the server's default when the user doesn't supply anything? I know in the drivers we're not supposed to send a write concern if the user doesn't supply one, but does that guidance extend to the tools?

>The string "majority" only exists on the left, delete code. Where are we testing that the default write concern is majority?

I've added an integration test to replace the deleted unit one that verifies the default behavior, and I did discover that in this patch, we are in fact ignoring write concern options set in the URI. I've made a slight (temporary) modification so that they aren't ignored any more, but that causes the default case to not send a write concern and use the MongoDB default instead (i.e. {w: 1}). Once we bump the go driver dependency, I can use the new options merging api to set "majority" as the forced default if that is what's desired.

Because we're no longer parsing the URI ourselves, the write concern information will be on the client/db/collection as set by the go driver rather than in our options types, hence why it must be an integration test.

Also, so that the arguments parsing is more easily testable, I added a `New` constructor function to `mongofiles` and introduced a new `Options` struct that encompasses all the different options types that can be used to configure mongofiles. This way, our integration tests can run by passing the same command-line arguments that a user would instead of manually constructing options types. I've found that the old integration tests combined with the unit tests are not really effective ways of testing the options parsing. The new test I added is an example of the new style. If you think it's the right approach, I will convert the rest of the integration tests, either in a later patch for this PR or in follow-up work. (I've already implemented a similar design in my wip [mongorestore port](https://github.com/patrickfreed/mongo-tools/blob/ffda33c3551f548bb6d924a82c4b30f1df4507f7/mongorestore/mongorestore_test.go#L91))

Reply:
The tool is documented as defaulting to majority, so I think we need to do that.